### PR TITLE
[new-lair-8] in-process keystore + ipc keystore

### PR DESCRIPTION
**Pull request by neonphog**
DEPENDS ON https://github.com/holochain/lair/pull/66

[Video Walkthrough (3:30)](https://www.youtube.com/watch?v=i3-bWkZj8sU)

Two server wrappers:
- one that uses `tokio::io::duplex` to facilitate an in-process keystore server, i.e. could be used for testing
- one that uses unix domain sockets on most systems or windows named-pipes on windows

### CODE REVIEW

File: crates/lair_keystore_api/src/ipc_keystore.rs [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+            // even if they tell us to do this backwards,
+            // let's at least try to do it correctly to start
+            match cli_hnd.hello(server_pub_key.clone()).await {
+                Ok(ver) => {
+                    priv_check_hello_ver(&opts, &ver)?;
+
+                    // hey, it worked, unlock and proceed
+                    cli_hnd.unlock(opts.passphrase.clone()).await?;
+                }
+                Err(e) if e.str_kind() == "KestoreLocked" => {
```
Comment (maackle) [inline crates/lair_keystore_api/src/ipc_keystore.rs:152]: typo?

**Review by maackle — APPROVED** [review]
Skimmed it with cursory understanding of the two implementations

File: crates/lair_keystore_api/src/ipc_keystore.rs [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+                    // lame, the server really is not unlocked...
+                    // unlock, but some attacker may get our passphrase
+                    cli_hnd.unlock(opts.passphrase.clone()).await?;
```
Comment (thedavidmeister) [inline crates/lair_keystore_api/src/ipc_keystore.rs:155]: given that the video says we're getting rid of this i think it's ok

**Review by thedavidmeister — APPROVED** [review]

File: crates/lair_keystore_api/src/ipc_keystore.rs (hunk 1 above)
Comment (neonphog) [reply crates/lair_keystore_api/src/ipc_keystore.rs:152]: yep! : ) - I've got a backlog task to normalize these kinds to an enum instead of strings...

**Pull request merged**